### ISSUE TITLE
bug: IonicReact multiple menu (different menu per tab ) not working  + direct navigating throws error Unhandled Rejection menu.componentOnReady not a func

### ISSUE THREAD

**Issue by ar-daniel** [ionic-team/ionic#21310]
# Bug Report

**Ionic version:**

[  ] **4.x**
[x] **5.x**

**Current behavior:**

- multiple menu - different menu per tab not working in `ionic-react`. (even after enabling the relevant menu using  `menuController ` in the life cycle hook. switching back to previous tab disables the menu / the menu button is disappeared.
- additionally implementing the  ` menuController.enable() ` has a side effect also - when loading the page directly instead of navigating through tabs  - getting an error - Unhandled Rejection (TypeError): menu.componentOnReady is not a function

**Expected behavior:**

- when switching back to previous tab the menu should load.
- page should load when directly navigating to page through url

**Steps to reproduce:**

**Related code:**

routes_page (containing the menu contents and IonRouterOutlet) is in separate from from the component that is rendered.

__in the routes_page.tsx__
```
// necessary imports here.

const route_Tools: React.FC = () => {
  return (
      
     <React.Fragment>
      <IonMenu menuId="toolsMenu" contentId="tools" id="idtools">
        <IonHeader>
          <IonToolbar>
            <IonTitle>Tools Menu</IonTitle>
          </IonToolbar>
        </IonHeader>
        <IonContent>
          <IonList>
            <IonMenuToggle menu="toolsMenu">
            <IonItem button routerLink="/tools/notepad" routerDirection="none">
                <IonIcon icon={create}></IonIcon>
                <IonLabel>Notepad</IonLabel>
            </IonItem>

            <IonItem button routerLink="/tools/calculator" routerDirection="none">
                <IonIcon icon={calculator}></IonIcon>
                <IonLabel>Calculator</IonLabel>
            </IonItem>

            </IonMenuToggle>
          </IonList>
        </IonContent>
      </IonMenu>


      <IonRouterOutlet id="tools">

        <Redirect path="/tools" to="/tools/home" exact />

        <Route path="/tools/home" >
          <Toolspage />
        </Route>
        <Route path="/tools/notepad" >
          {/* <Tool_notepad /> */}
        </Route>
        <Route path="/tools/calculator" >
          {/* <Tool_calculator /> */}
        </Route>

      </IonRouterOutlet>
      </React.Fragment>

  )
};
```

__And in the component page tsx which is rendered.__
menuController.enable(true,“menuId”) is used within useIonViewWillEnter() method to execute when ever the page is about to load.

```
//other imports before as necessary
import { menuController } from '@ionic/core';

const ToolsPage: React.FC = () => {

  useIonViewWillEnter(async() => {
    await menuController.enable(true,"toolsMenu")
  })

  useIonViewWillLeave(async() => {
    await menuController.enable(false,"toolsMenu")
  })

  return (
    <IonPage>
        <IonHeader>
          <IonToolbar>
            <IonButtons slot="start">
              <IonMenuButton menu="toolsMenu" ></IonMenuButton>
            </IonButtons>
            <IonTitle>Tools Menu page</IonTitle>
          </IonToolbar>
        </IonHeader>

      <IonContent>
{/* page content here */}
      </IonContent>
    </IonPage>
  );
};

export default ToolsPage;
```
similar setup done for the other tab page also.

**Other information:**

Enabling the particular menu for every page was a fix but I guess it works only for angular and NOT for react ( saw in a video) https://www.youtube.com/watch?v=4VYC725VMNA
watch from ( 12:10 to 16:30 ) for problem and work around. 

when checking with menuController.isEnabled() - the particular menu does get enabled (true value set) and the others disabled.
hence the correct menu gets enabled by life cycle hook  but still somehow doesn't work

**Ionic info:**

```
Ionic:
   Ionic CLI       : 6.8.0 (/usr/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 5.1.0

Capacitor:
   Capacitor CLI   : 2.1.0
   @capacitor/core : 2.1.0

Utility:
   cordova-res : not installed
   native-run  : not installed

System:
   NodeJS : v14.2.0 (/usr/bin/node)
   npm    : 6.14.4
   OS     : Linux 5.0
```

**Comment by ar-daniel** [comment]
Ok with a lot of work - created a standalone code sample to replicate the issue in an online code editor. Links below.

https://stackblitz.com/edit/ionic-react-ts-multiple-menu-per-tab-issue?file=src%2FApp.tsx
https://ionic-react-ts-multiple-menu-per-tab-issue.stackblitz.io/home

the second issue does'nt seem to show up in stackblitz. Please download a copy in your local machine and try running. 

By the way, Does stackblitz have a Ionic-React workspace. ? replicating a ionic-react-ts from an react-ts workspace was quite a pain and time consuming.

**Comment by liamdebeasi** [comment]
Thanks for the follow up. Can you update this reproduction to not use both `@ionic/react` and `@ionic/core`? Using both at the same time can sometimes cause issues. If you need to access a menu, you can pass it via a prop to each page rather than importing the menuController from `@ionic/core`.

**Comment by ar-daniel** [comment]
I'm not getting it - 'sending the menu via a prop to each page.'

I could have the IonMenu in a separate component and import it into the page that needs it. But the menu works through a IonMenuButton with the attribute menu="set to menu ID of the menu". 
We couldn't just place the imported menu itself there instead of the button. It didn't work though. 

```

// import { menuController } from '@ionic/core';
import Tab1PageMenu from './Tab1PageMenu'

const Tab1page: React.FC = () => {

  // useIonViewWillEnter(async () => {
  //   await menuController.enable(true, "tab1Menu")
  // })

  // useIonViewWillLeave(async () => {
  //   await menuController.enable(false, "tab1Menu")
  // })

  return (
    <React.Fragment>
      <IonPage>
        <IonHeader>
          <IonToolbar>
            <Tab1PageMenu />
            {/* <IonButtons slot="start">
            <IonMenuButton menu="tab1Menu" >
            </IonMenuButton>
          </IonButtons> */}
            <IonTitle>Tab 1</IonTitle>
          </IonToolbar>
        </IonHeader>
        <IonContent>
// ...

         </IonContent>
      </IonPage>
    </React.Fragment>
  );
};
```

Please clarify. Do you have any example/ reference code for this.?

**Comment by liamdebeasi** [comment]
Sorry, my previous comment was vague. What I mean is rather than import `menuController` via @ionic/core, you should try modifying the properties on the menu itself. For example, when you want to disable a menu, you should set the `disabled` property on `IonMenu` rather than trying to do it via the controller. The `menuController` import in Ionic Core is really intended for vanilla JS applications, not for React apps.

Check out our Ionic React example app: https://github.com/ionic-team/ionic-react-conference-app/blob/master/src/components/Menu.tsx. We are setting the `disabled` property on `IonMenu` based upon the value of `menuEnabled` which we have connected to our state. I would recommend trying something like that to see if it works properly that way.

**Comment by ar-daniel** [comment]
I've made the revision as advised and spent lot of time researching  where and all its failing.
The updated code using react context api for storing & updating the menu enable property and using it in each of Menu's disable property is in link below. ( old code kept in src_menucontroller)

https://stackblitz.com/edit/ionic-react-ts-multiple-menu-per-tab-issue?file=src%2FTab1Page.tsx
https://ionic-react-ts-multiple-menu-per-tab-issue.stackblitz.io/tab1/home

__Inference :__
Its NOT working -
   - the context store is working though ( as seen from the menu state values in each page.)
   - hard coding the menu's disable attribute values also did not work ( eg keeping the disable attribute of tabmenu as false and other menu's as true )
   -  going one step further - removing the other menus and keeping (importing & using) only one menu, say Tab2PageMenu. - didn't work when coming back to the page.

when does it work -
 - when there is only one menu, and page that renders the menu is directly called from the react function where IonTabs/ IonTabButtons are declared. (not in a nested route ).

Hence,
 - __associate issue__  - if the the component that renders the menu is called from a nested route  from the react function where IonTabs/ IonTabButtons is declared. - menu does'nt render on revisiting the page. (which is how our app is written)

Please __help resolve the issue to successfully render different menu per page__ and ideally without needing to use context or redux. Maintaining, accessing, updating a store variable for rendering menus is lot of additional work. This need to be baked into the framework as its part of the rendering the UI.

**Comment by ar-daniel** [comment]
@liamdebeasi - Is there an update. ? Were you able to find/ fix the problem causing this issue.?

**Comment by liamdebeasi** [comment]
Hey there,

Apologies for the delay. I discussed this with the team and we concluded this is likely due to how your application is structured, not due to a bug in Ionic Framework.

I created a simplified version of your multi-menu setup that required less state management and did not result in any of the errors you reported: https://github.com/liamdebeasi/ionic-react-multi-menu. Feel free to clone this and use this setup in your application.

That being said, we advise against having multiple instances of the menu. It likely would be easier to have one instance of a menu and swap out the menu's content based upon the tab that is active.

Thanks!
